doc/code/curves.md

```diff
@@ -0,0 +1,217 @@
+# Curves
+
+*Curves* are data structures that manage changes to data over time. More precisely,
+they do not only store data but time-value pairs (keyframes) that allow the retrieval
+of a curve data value for any point in time.
+
+Curves are an integral part of openage's event-based game simulation.
+
+1. [Motivation](#motivation)
+2. [Architecture](#architecture)
+3. [Curve Types](#curve-types)
+   1. [Primitive](#primitive)
+      1. [Common Operations](#common-operations)
+      2. [Discrete](#discrete)
+      3. [Continuous](#continuous)
+      4. [Segmented](#segmented)
+   2. [Container](#container)
+      1. [Queue](#queue)
+      2. [Unordered Map](#unordered-map)
+
+
+## Motivation
+
+The openage curve structures are inspired by a [similar implementation](https://www.forrestthewoods.com/blog/tech_of_planetary_annihilation_chrono_cam/#.lmxbu3vld) from the game Planetary Annihilation.
+
+Curves intend to solve synchronicity problems in games. In traditional implementations
+like lockstep, data is changed incrementally in regular intervals (ticks). Keeping
+these changes and the overall gamestate consistent/in sync over network or across threads
+is very important, as one missing change can result in a desync of the entire simulation.
+Recovering from a desync can also be very hard.
+
+In comparison, simulation with curves allows both async operations and an easier recovery
+from desync states. Data is not changed incrementally but is instead calculated using
+*keyframe interpolation*. In other words, changes to data are treated as keyframes on
+a timeline with in-between values being interpolated from adjacent keyframes.
+
+![Curve keyframes example](images/continuous_curve.png)
+
+Additionally, curves cannot only access values for the current simulation time but also for
+any past or even future times. Keyframes can also be inserted for any point in time without
+directly invalidating the state, making curves more reliable in async scenarios (although
+resolving dependencies for keyframes in the past can still be challenging).
+
+The usage of curves has a few downsides though. They are less space efficient due to the
+keyframe storage, interpolation are more costly more costly than incremental changes, and
+their integration is more complex than the usage of simpler data structures. However, in
+situations where operations are predictable, long-lasting, and easy to calculate - which
+is the case for most RTS games - these downsides are usually outweight by the positives.
```

Comment:
usually? :D i'd rather say "probably"

Reply:
hey I'm trying to sell something here :D

(rephrased it nevertheless)